buildroot/share/scripts/upload.py

```diff
@@ -0,0 +1,193 @@
+# Upload Callback #
+#-----------------#
+def Upload(source, target, env):
+
+    #----------------#
+    # Port functions #
+    #----------------#
+    def _GetUploadPort(env):
+        if Debug: print("Autodetecting upload port...")
+        env.AutodetectUploadPort(env)
+        port = env.subst('$UPLOAD_PORT')
+        if not port:
+            print("Error detecting the upload port.")
+            return None
+        if Debug: print('OK')
+        return port
+
+    #-------------------------#
+    # Simple serial functions #
+    #-------------------------#
+    def _Send(data):
+        if Debug: print(">> " + data)
+        strdata = bytearray(data, "utf8") + b'\n'
+        port.write(strdata)
+        time.sleep(0.010)
+
+    def _Recv():
+        clean_responses = []
+        responses = port.readlines()
+        for Resp in responses:
+            clean_response = Resp.decode('utf8').rstrip().lstrip()
+            clean_responses.append(clean_response)
+            if Debug: print("<< " + clean_response)
+        return clean_responses
+        
+    #------------------#
+    # SDCard functions #
+    #------------------#
+    def _CheckSDCard():
+        if Debug: print('Checking SD card...')
+        _Send("M21")
+        Responses = _Recv()
+        if len(Responses) < 1 or not 'SD card ok' in Responses[0]:
+            print("Error accessing SD card")
+            return False
+        if Debug: print('SD Card OK')
+        return True
+
+    #----------------#
+    # File functions #
+    #----------------#
+    def _GetFirmwareFiles():
+        if Debug: print('Get firmware files...')
+        _Send("M20 F")
+        Responses = _Recv()
+        if len(Responses) < 3 or not 'file list' in Responses[0]:
+            print("Error getting firmware files")
+            return (False, None)
+        if Debug: print('OK')
+        return (True, Responses)
+        
+    def _FilterFirmwareFiles(FirmwareList):
+        Firmwares = []
+        for FWFile in FirmwareList:
+            if not "/" in FWFile and ".BIN" in FWFile:
+                idx = FWFile.index(".BIN")
+                Firmwares.append(FWFile[:idx+4])
+        return Firmwares
+
+    def _RemoveFirmwareFile(FWFile):
+        _Send("M30 /" + FWFile)
+        Responses = _Recv()
+        return len(Responses) >= 1 and 'File deleted' in Responses[0]
+
+
+    #---------------------#
+    # Callback Entrypoint #
+    #---------------------#
+    # Get firmware upload params
+    upload_firmware_source_name = str(source[0])    # Source firmware filename
+    upload_speed = env['UPLOAD_SPEED'] if 'UPLOAD_SPEED' in env else 115200
+                                                    # baud rate of serial connection
+    upload_port = _GetUploadPort(env)               # Serial port to use
+    if not upload_port: exit(1)
+    # Set local upload params
+    upload_firmware_target_name = 'fw-' + ''.join(random.choices("ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789", k=5)) + ".BIN"
```

Comment:
Does the build not already produce a properly "name mangled" file? I thought it appended a timestamp to the filename or something, to make sure that it was always unique. Is this a workaround to deal with 8.3 naming?

Reply:
We would certainly want this aspect to be optional, since most boards just want "firmware.bin" or some other fixed filename.

I personally think it could be acceptable to put it in initially working with one popular board as you have then rework it to be more generic in another pass.

Reply:
The 8.3 file format was needed in the initial (manual) scripting due to problems with the tilde~ character into VSCode serial monitor, but I think it can be removed using the firmware name.
Also, if all the firmware names are correctly generated by pio for every board, we don't need to do anything else...although an override for the destination filename can be useful.